homeassistant/components/tplink/__init__.py

```diff
@@ -86,11 +114,43 @@ async def _async_discovery(*_: Any) -> None:
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up TPLink from a config entry."""
     host = entry.data[CONF_HOST]
+
+    credentials = await get_stored_credentials(hass)
+    connection_params = None
+
+    if connection_params_dict := entry.data.get(CONF_CONNECTION_PARAMS):
+        try:
+            connection_params = ConnectionParameters.from_dict(connection_params_dict)
+        except SmartDeviceException:
+            _LOGGER.warning(
+                "Invalid connection parameters for %s: %s", host, connection_params_dict
+            )
     try:
-        device: SmartDevice = await Discover.discover_single(host, timeout=10)
+        httpx_asyncclient = create_async_httpx_client(hass, verify_ssl=False)
+        device: SmartDevice = await Discover.connect_single(
+            host,
+            timeout=10,
+            credentials=credentials,
+            connection_params=connection_params,
+            httpx_asyncclient=httpx_asyncclient,
+            try_discovery_on_error=True,
+        )
```

Comment:
If we would store all connection parameters in the object, this would be simplified to be something like:
```
conn_params = ConnectionParameters.from_dict(connection_params_dict)
conn_params.httpx_client = httpx_asyncclient
conn_params.try_discovery_on_error = True  # I'm unsure if we need to try a discovery on errors, but let the auto-discovery do its trick?
device = SmartDevice.connect(connection_params)
```

Reply:
Done